### DISCUSSION TITLE
Miner Adding storage power

### DISCUSSION THREAD

**Discussion by JaisPiyush** [JaisPiyush/arc-ip#2] [Ideas]
## Summary
The protocol allows undercollateralized lending to Filecoin Node operators. Each node contributes storage in the chain and utilization of contributed storage accumulates fees. The fees are the active income on the chain for node operators, more about the economics can be read here [Filecoin SP](https://lotus.filecoin.io/storage-providers/get-started/overview/).

Anyone with enough storage can participate in the chain but to win client deals each node must provide collateral (guarantee against the storage). Undercollateralized lending helps nodes to raise initial collateral funds.

### Appendix
1. **Beneficiary Address**: A beneficiary address is a special purpose address that takes over the financial control from the owner address for a given quota and time, but does not take over the owner addresses privileges. The separation of owner and beneficiary brings more flexibility to the Filecoin financing market. [Addresses](https://lotus.filecoin.io/storage-providers/operate/addresses/#beneficiary-address).

2. **Sector**: Sectors are the basic units of storage on Filecoin. They have standard sizes, as well as well-defined time increments for commitments. The size of a sector balances security concerns against usability. A sector lifetime is determined in the storage market and sets the promised duration of the sector. [Sector](https://spec.filecoin.io/systems/filecoin_mining/sector/)

3. **Oracles**: Oracles are responsible for syncing off-chain data to on-chain and performing the expensive task off-chain with proof.

#### Prerequisites
1. Beneficiary Address and Filecoin miner address
2. Sector
4. Oracle

#### Structs
``` solidity
 
struct AddingPowerRequest {
    uint256 rawBytePower;
    uint32 commitmentTime;
    uint256 maxPledgeAmount;
    uint256 sectorSize;  
    address miner;
    // owner of the oracle must be a miner if the oracle already exists
    address oracle; 

}

enum StakePoolStatus {
    // Stake Pool is in the collateral queue
    InQueue,
    // Stake pool is waiting for the node to transfer the onwerhsip address to StakePool
    // with updated quota and epoch
    OwnershipConfirmationPending,
    // Stake Pool is waiting for the node to host an oracle and confirm its liveliness by responding to a challenge
    OracleConfirmationPending,
    // Stake Pool is active
    Active,
    // Stake Pool is paused
    Paused,
    // Stake Pool is terminated
    Terminated,
     // Stake pool is active but the miner is in the queue to add more power
    AsyncInQueue,

}



struct PendingCollateral {
    address stakePool;
    AddingPowerRequest request;
    uint32 addedInQueueTimestamp;
    uint32 proposalExpireTimestamp;
    bool isNewStakePool;
}

      
```

<img width="3200" alt="Lending Protocol Adding Collateral" src="https://user-images.githubusercontent.com/46999854/223335716-88dc5386-17ff-4714-97e1-36d53019a881.png">

* The node operator submits the request to join the pool with `AddingPowerRequest` as a parameter.

*  The request `PendingCollateral` will be added to the collateral queue (in Staking Pool Manager smart contract).
   - Before adding the collateral to the queue the contract will check the existence of StakePool related to the miner. 
        - If no Stakepool exists one will be created immediately with stake pool status `StakePoolStatus.InQueue`.
        - If Stakepool exists the status will be changed to `StakePoolStatus.AsyncInQueue`.
   - `maxPledgeAmount` is only the initial pledge collateral and without the sealing cost. Sealing of the sector will be covered by the miner's balance.
* The oracle will optimistically pick the best stake pools to invest in at set intervals (24 hrs). 
    - Existing stake pools will be given priority.
    - New stake pools will be shorted based on storage power and commitment time against available funds in the protocol.
    - The selection of nodes must be deterministic and provable by 51% Oracle DAO consent.
    - The oracle address will be added to the whitelist.
* The selected StakePool:
    - Suitable APR will be calculated according to the locked collateral and storage power. (APR will be auto-updated when new collateral is added).
    - If the miner's owner is different than Stake Pool then initiate ownership address change with hard deadline of 3 days.
        - if miner fails to change the owner address before the deadline, Fees from the miner's locked amount will be deducted, residual amounts will be refunded, Stake Pool will be destroyed.
        - else Initiate oracle confirmation with another deadline of 3 days.
* The oracle associated with the node must be confirmed using a challenge.
    - Oracle address for the node can be changed only by the current oracle address.
    - A functionality should be provided to confirm the activeness of an oracle.
* After ownership and oracle setup, the amounts will be pulled from the vault into the StakePool and wait for the node operator to publish `sector pledge` command from the lotus cli (deadline 1 week).

**Comment by hcheng826** [comment]
1. Is the relation like one miner will have one Stake Pool, and each Stake Pool will have one Oracle? Or it's one universal Oracle serving all the miners?
2. I saw Oracle DAO, so our oracle is controlled by a governance mechanism?
3. I think for Oracle for convention in Defi applications, it's maintaining the off-chain data to be available on-chain. But it don't actually interact with other contracts or changing states of other contracts. Is there a different meaning for Oracle under the context of filecoin?

**Reply by JaisPiyush** [reply]
1. Yes, each miner node will have particular stake pool and oracle.
2. Oracle DAO will govern the validity of data, oracles that can contribute in data feeding.
3. Each oracle will feed off-chain data and 51% quorum will make the data valid on-chain. Then each individual oracle will claim rewards for their node.
4. Yes, these oracles are basically doing what chainlink does bts. Uploads the data on-chain so other contracts can utilize them. It also works as job scheduler and cron jobber. This is also required bcz of limitations of evm smart contracts to perform expensive tasks

**Reply by JaisPiyush** [reply]
Oracle DAO will also govern the registration/removal of oracles from the network. Unregistwred oracles won't be able to feed data or claim rewards.

**Comment by JaisPiyush** [comment]
1. The protocol should have min. Storage power threshold (3PiB for mainnet).
2. Min. commitment time

**Comment by JaisPiyush** [comment]
And, we should also have proper sytem to mock the entire process for scaling ethereum hackathon.
